kex-runner/src/main/kotlin/org/vorpal/research/kex/asm/analysis/symbolic/SymbolicPathSelector.kt

```diff
@@ -21,3 +25,85 @@ class DequePathSelector : SymbolicPathSelector {
 
     override suspend fun next(): Pair<TraverserState, BasicBlock> = queue.poll()
 }
+
+/**
+ * Implements n-subpath algorithm, which makes path decisions based on the frequency of visits to paths of length n
+ */
+class NSubpathPathSelector(val n: Int = 2) : SymbolicPathSelector {
+    private val eSVector = mutableSetOf<ExecutionState>()
+    private val pathVisits = mutableMapOf<List<BasicBlock>, Int>()
+
+    override suspend fun add(state: TraverserState, block: BasicBlock) {
+        val newState = ExecutionState(state to block, n)
+        eSVector.add(newState)
+        if (pathVisits[newState.path] == null) {
+            pathVisits[newState.path] = 0
+        }
+    }
+
+    override suspend fun hasNext(): Boolean = eSVector.isNotEmpty()
+
+    override suspend fun next(): Pair<TraverserState, BasicBlock> {
+        var minPath = Int.MAX_VALUE
+        eSVector.forEach { minPath = min(minPath, pathVisits[it.path]!!) }
+        val selectSet = mutableSetOf<ExecutionState>()
+
+        for (elem in eSVector) {
+            if (pathVisits[elem.path] == minPath) {
+                selectSet.add(elem)
+            }
+        }
+
+        val result =  selectSet.random()
+        pathVisits[result.path] = pathVisits[result.path]!! + 1
+        eSVector.remove(result)
+        return result.node
+    }
+
+    class ExecutionState(
+        val node: Pair<TraverserState, BasicBlock>,
+        pathLength: Int
+    ) {
+        val path: List<BasicBlock> = node.first.blockPath.takeLast(pathLength)
```

Comment:
Our algorithm currently only works on a basic block level, but it may be too broad for us, because each basic block may contain several branching points: method calls, null-checks, branch/jump instructions. I'm not sure that BB level is enough here

Reply:
Although, I am not sure that it will really affect the performance. We will need to discuss it later